File: domain/src/main/java/co/joebirch/bastion/domain/interactor/UseCase.java
package co.joebirch.bastion.domain.interactor;

import co.joebirch.bastion.domain.executor.PostExecutionThread;
import co.joebirch.bastion.domain.executor.ThreadExecutor;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

public abstract class UseCase<T, Params> {

    private final ThreadExecutor threadExecutor;
    private final PostExecutionThread postExecutionThread;


    protected UseCase(ThreadExecutor threadExecutor,
                      PostExecutionThread postExecutionThread) {
        this.threadExecutor = threadExecutor;
        this.postExecutionThread = postExecutionThread;
    }

    /**
     * Builds an {@link Single} which will be used when executing the current {@link UseCase}.
     */
    abstract Single<T> buildUseCaseObservable(Params params);

    /**
     * Executes the current use case.
     */
    @SuppressWarnings("unchecked")
    public Single<T> execute(Params params) {
        return this.buildUseCaseObservable(params)
                .subscribeOn(Schedulers.from(threadExecutor))
                .observeOn(postExecutionThread.getScheduler());
    }

}

File: domain/src/main/java/co/joebirch/bastion/domain/interactor/UseCase.kt
package co.joebirch.bastion.domain.interactor

import co.joebirch.bastion.domain.executor.PostExecutionThread
import co.joebirch.bastion.domain.executor.ThreadExecutor
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

abstract class UseCase<T, in Params>(private val postExecutionThread: PostExecutionThread,
                                     private val threadExecutor: ThreadExecutor) {

    abstract fun buildUseCaseObservable(params: Params?): Single<T>

    fun execute(params: Params?): Single<T> {
        return buildUseCaseObservable(params)
                .subscribeOn(Schedulers.from(threadExecutor))
                .observeOn(postExecutionThread.getScheduler())
    }

}